File: backend/src/handlers/transactions.rs
// Copyright (c) 2024 Tempo Budget
// SPDX-License-Identifier: MIT
//
// Transaction management HTTP handlers.

//! # Transaction Handlers
//!
//! This module provides HTTP handlers for transaction management:
//! - `GET /api/budgets/:budget_id/transactions` - List transactions
//! - `POST /api/budgets/:budget_id/transactions` - Create a transaction
//! - `DELETE /api/transactions/:id` - Delete a transaction
//!
//! Also includes recurring transaction handlers:
//! - `GET /api/budgets/:budget_id/recurring` - List recurring transactions
//! - `POST /api/budgets/:budget_id/recurring` - Create a recurring transaction
//! - `PUT /api/recurring/:id/toggle` - Toggle recurring transaction active status
//! - `DELETE /api/recurring/:id` - Delete a recurring transaction

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use std::sync::Arc;
use uuid::Uuid;
use chrono::Utc;
use utoipa;

use crate::{
    models::{
        CreateRecurringTransaction, CreateTransaction, RecurringTransaction, Transaction,
    },
    DbPool,
};

/// Retrieves all transactions for a specific budget.
///
/// # Endpoint
/// `GET /api/budgets/:budget_id/transactions`
///
/// # Path Parameters
/// - `budget_id`: The budget's unique identifier
///
/// # Returns
/// - `200 OK` with array of `Transaction` objects (sorted by date descending)
/// - `500 Internal Server Error` if the query fails
#[utoipa::path(
    get,
    path = "/api/budgets/{budget_id}/transactions",
    tag = "transactions",
    params(
        ("budget_id" = String, Path, description = "Budget unique identifier")
    ),
    responses(
        (status = 200, description = "List of transactions", body = Vec<Transaction>),
        (status = 500, description = "Failed to fetch transactions")
    )
)]
pub async fn get_transactions(
    State(pool): State<Arc<DbPool>>,
    Path(budget_id): Path<String>,
) -> Result<Json<Vec<Transaction>>, StatusCode> {
    let transactions = sqlx::query_as::<_, Transaction>(
        "SELECT id, budget_id, category_id, title, amount, transaction_type, date, comment,
         is_recurring, created_at FROM transactions WHERE budget_id = ? ORDER BY date DESC"
    )
        .bind(&budget_id)
        .fetch_all(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(Json(transactions))
}

/// Creates a new transaction.
///
/// # Endpoint
/// `POST /api/budgets/:budget_id/transactions`
///
/// # Request Body
/// - `budget_id`: Budget ID for the transaction
/// - `category_id`: Category ID for the transaction
/// - `title`: Transaction title/description
/// - `amount`: Transaction amount
/// - `transaction_type`: Either "income" or "expense"
/// - `date`: Transaction date (ISO 8601 format)
/// - `comment`: (optional) Additional notes
///
/// # Returns
/// - `200 OK` with the created `Transaction` object
/// - `500 Internal Server Error` if creation fails
#[utoipa::path(
    post,
    path = "/api/budgets/{budget_id}/transactions",
    tag = "transactions",
    request_body = CreateTransaction,
    responses(
        (status = 200, description = "Transaction created successfully", body = Transaction),
        (status = 500, description = "Failed to create transaction")
    )
)]
pub async fn create_transaction(
    State(pool): State<Arc<DbPool>>,
    Json(payload): Json<CreateTransaction>,
) -> Result<Json<Transaction>, StatusCode> {
    let id = Uuid::new_v4().to_string();
    let now = Utc::now().to_rfc3339();

    sqlx::query(
        "INSERT INTO transactions (id, budget_id, category_id, title, amount, transaction_type,
         date, comment, is_recurring, created_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, 0, ?)"
    )
        .bind(&id)
        .bind(&payload.budget_id)
        .bind(&payload.category_id)
        .bind(&payload.title)
        .bind(payload.amount)
        .bind(&payload.transaction_type)
        .bind(&payload.date)
        .bind(&payload.comment)
        .bind(&now)
        .execute(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    let transaction = sqlx::query_as::<_, Transaction>(
        "SELECT id, budget_id, category_id, title, amount, transaction_type, date, comment,
         is_recurring, created_at FROM transactions WHERE id = ?"
    )
        .bind(&id)
        .fetch_one(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(Json(transaction))
}

/// Deletes a transaction.
///
/// # Endpoint
/// `DELETE /api/transactions/:id`
///
/// # Path Parameters
/// - `id`: The transaction's unique identifier
///
/// # Returns
/// - `204 No Content` on successful deletion
/// - `500 Internal Server Error` if deletion fails
#[utoipa::path(
    delete,
    path = "/api/transactions/{id}",
    tag = "transactions",
    params(
        ("id" = String, Path, description = "Transaction unique identifier")
    ),
    responses(
        (status = 204, description = "Transaction deleted successfully"),
        (status = 500, description = "Failed to delete transaction")
    )
)]
pub async fn delete_transaction(
    State(pool): State<Arc<DbPool>>,
    Path(id): Path<String>,
) -> Result<StatusCode, StatusCode> {
    sqlx::query("DELETE FROM transactions WHERE id = ?")
        .bind(&id)
        .execute(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(StatusCode::NO_CONTENT)
}

/// Retrieves all recurring transactions for a specific budget.
///
/// # Endpoint
/// `GET /api/budgets/:budget_id/recurring`
///
/// # Path Parameters
/// - `budget_id`: The budget's unique identifier
///
/// # Returns
/// - `200 OK` with array of `RecurringTransaction` objects
/// - `500 Internal Server Error` if the query fails
#[utoipa::path(
    get,
    path = "/api/budgets/{budget_id}/recurring",
    tag = "transactions",
    params(
        ("budget_id" = String, Path, description = "Budget unique identifier")
    ),
    responses(
        (status = 200, description = "List of recurring transactions", body = Vec<RecurringTransaction>),
        (status = 500, description = "Failed to fetch recurring transactions")
    )
)]
pub async fn get_recurring_transactions(
    State(pool): State<Arc<DbPool>>,
    Path(budget_id): Path<String>,
) -> Result<Json<Vec<RecurringTransaction>>, StatusCode> {
    let transactions = sqlx::query_as::<_, RecurringTransaction>(
        "SELECT id, budget_id, category_id, title, amount, transaction_type, frequency, day,
         active, created_at FROM recurring_transactions WHERE budget_id = ?"
    )
        .bind(&budget_id)
        .fetch_all(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(Json(transactions))
}

/// Creates a new recurring transaction.
///
/// # Endpoint
/// `POST /api/budgets/:budget_id/recurring`
///
/// # Request Body
/// - `budget_id`: Budget ID for the recurring transaction
/// - `category_id`: Category ID for generated transactions
/// - `title`: Transaction title/description
/// - `amount`: Transaction amount
/// - `transaction_type`: Either "income" or "expense"
/// - `frequency`: One of "daily", "weekly", "monthly", "yearly"
/// - `day`: (optional) Day of the period for scheduling
///
/// # Returns
/// - `200 OK` with the created `RecurringTransaction` object
/// - `500 Internal Server Error` if creation fails
#[utoipa::path(
    post,
    path = "/api/budgets/{budget_id}/recurring",
    tag = "transactions",
    request_body = CreateRecurringTransaction,
    responses(
        (status = 200, description = "Recurring transaction created successfully", body = RecurringTransaction),
        (status = 500, description = "Failed to create recurring transaction")
    )
)]
pub async fn create_recurring_transaction(
    State(pool): State<Arc<DbPool>>,
    Json(payload): Json<CreateRecurringTransaction>,
) -> Result<Json<RecurringTransaction>, StatusCode> {
    let id = Uuid::new_v4().to_string();
    let now = Utc::now().to_rfc3339();

    sqlx::query(
        "INSERT INTO recurring_transactions (id, budget_id, category_id, title, amount,
         transaction_type, frequency, day, active, created_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, 1, ?)"
    )
        .bind(&id)
        .bind(&payload.budget_id)
        .bind(&payload.category_id)
        .bind(&payload.title)
        .bind(payload.amount)
        .bind(&payload.transaction_type)
        .bind(&payload.frequency)
        .bind(payload.day)
        .bind(&now)
        .execute(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    let transaction = sqlx::query_as::<_, RecurringTransaction>(
        "SELECT id, budget_id, category_id, title, amount, transaction_type, frequency, day,
         active, created_at FROM recurring_transactions WHERE id = ?"
    )
        .bind(&id)
        .fetch_one(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(Json(transaction))
}

/// Toggles the active status of a recurring transaction.
///
/// # Endpoint
/// `PUT /api/recurring/:id/toggle`
///
/// # Path Parameters
/// - `id`: The recurring transaction's unique identifier
///
/// # Returns
/// - `200 OK` with the updated `RecurringTransaction` object
/// - `404 Not Found` if the recurring transaction doesn't exist
/// - `500 Internal Server Error` if update fails
#[utoipa::path(
    put,
    path = "/api/recurring/{id}/toggle",
    tag = "transactions",
    params(
        ("id" = String, Path, description = "Recurring transaction unique identifier")
    ),
    responses(
        (status = 200, description = "Recurring transaction toggled successfully", body = RecurringTransaction),
        (status = 404, description = "Recurring transaction not found"),
        (status = 500, description = "Failed to toggle recurring transaction")
    )
)]
pub async fn toggle_recurring_transaction(
    State(pool): State<Arc<DbPool>>,
    Path(id): Path<String>,
) -> Result<Json<RecurringTransaction>, StatusCode> {
    // Get the current state
    let current = sqlx::query_as::<_, RecurringTransaction>(
        "SELECT id, budget_id, category_id, title, amount, transaction_type, frequency, day,
         active, created_at FROM recurring_transactions WHERE id = ?"
    )
        .bind(&id)
        .fetch_one(pool.as_ref())
        .await
        .map_err(|_| StatusCode::NOT_FOUND)?;

    // Toggle the state
    let new_active = !current.active;

    sqlx::query("UPDATE recurring_transactions SET active = ? WHERE id = ?")
        .bind(new_active)
        .bind(&id)
        .execute(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    // Get the updated version
    let updated = sqlx::query_as::<_, RecurringTransaction>(
        "SELECT id, budget_id, category_id, title, amount, transaction_type, frequency, day,
         active, created_at FROM recurring_transactions WHERE id = ?"
    )
        .bind(&id)
        .fetch_one(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(Json(updated))
}

/// Deletes a recurring transaction.
///
/// # Endpoint
/// `DELETE /api/recurring/:id`
///
/// # Path Parameters
/// - `id`: The recurring transaction's unique identifier
///
/// # Returns
/// - `204 No Content` on successful deletion
/// - `500 Internal Server Error` if deletion fails
#[utoipa::path(
    delete,
    path = "/api/recurring/{id}",
    tag = "transactions",
    params(
        ("id" = String, Path, description = "Recurring transaction unique identifier")
    ),
    responses(
        (status = 204, description = "Recurring transaction deleted successfully"),
        (status = 500, description = "Failed to delete recurring transaction")
    )
)]
pub async fn delete_recurring_transaction(
    State(pool): State<Arc<DbPool>>,
    Path(id): Path<String>,
) -> Result<StatusCode, StatusCode> {
    sqlx::query("DELETE FROM recurring_transactions WHERE id = ?")
        .bind(&id)
        .execute(pool.as_ref())
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(StatusCode::NO_CONTENT)
}

File: backend/app/routes/transactions.py
# Copyright (c) 2024 Tempo Budget
# SPDX-License-Identifier: MIT
"""Transaction management routes."""

from datetime import datetime, timezone
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from ..database import get_db
from ..models import (
    Transaction, CreateTransaction,
    RecurringTransaction, CreateRecurringTransaction
)

router = APIRouter()


@router.get("/budgets/{budget_id}/transactions", response_model=list[Transaction])
async def get_transactions(budget_id: str, db: AsyncSession = Depends(get_db)):
    """Retrieve all transactions for a specific budget."""
    result = await db.execute(
        text("""
            SELECT id, budget_id, category_id, title, amount, transaction_type,
                   date, comment, is_recurring, paid_by_user_id, created_at
            FROM transactions WHERE budget_id = :budget_id ORDER BY date DESC
        """),
        {"budget_id": budget_id}
    )
    rows = result.fetchall()
    return [Transaction(
        id=row.id,
        budget_id=row.budget_id,
        category_id=row.category_id,
        title=row.title,
        amount=row.amount,
        transaction_type=row.transaction_type,
        date=row.date,
        comment=row.comment,
        is_recurring=row.is_recurring,
        paid_by_user_id=row.paid_by_user_id,
        created_at=row.created_at,
    ) for row in rows]


@router.post("/budgets/{budget_id}/transactions", response_model=Transaction)
async def create_transaction(
    budget_id: str,
    payload: CreateTransaction,
    db: AsyncSession = Depends(get_db)
):
    """Create a new transaction."""
    transaction_id = str(uuid4())
    now = datetime.now(timezone.utc).isoformat()

    await db.execute(
        text("""
            INSERT INTO transactions (id, budget_id, category_id, title, amount,
                                       transaction_type, date, comment, is_recurring, paid_by_user_id, created_at)
            VALUES (:id, :budget_id, :category_id, :title, :amount,
                    :transaction_type, :date, :comment, 0, :paid_by_user_id, :created_at)
        """),
        {
            "id": transaction_id,
            "budget_id": budget_id,
            "category_id": payload.category_id,
            "title": payload.title,
            "amount": payload.amount,
            "transaction_type": payload.transaction_type,
            "date": payload.date,
            "comment": payload.comment,
            "paid_by_user_id": payload.paid_by_user_id,
            "created_at": now,
        }
    )
    await db.commit()

    result = await db.execute(
        text("""
            SELECT id, budget_id, category_id, title, amount, transaction_type,
                   date, comment, is_recurring, paid_by_user_id, created_at
            FROM transactions WHERE id = :id
        """),
        {"id": transaction_id}
    )
    row = result.fetchone()
    return Transaction(
        id=row.id,
        budget_id=row.budget_id,
        category_id=row.category_id,
        title=row.title,
        amount=row.amount,
        transaction_type=row.transaction_type,
        date=row.date,
        comment=row.comment,
        is_recurring=row.is_recurring,
        paid_by_user_id=row.paid_by_user_id,
        created_at=row.created_at,
    )


@router.delete("/transactions/{transaction_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_transaction(transaction_id: str, db: AsyncSession = Depends(get_db)):
    """Delete a transaction."""
    await db.execute(text("DELETE FROM transactions WHERE id = :id"), {"id": transaction_id})
    await db.commit()


@router.get("/budgets/{budget_id}/recurring", response_model=list[RecurringTransaction])
async def get_recurring_transactions(budget_id: str, db: AsyncSession = Depends(get_db)):
    """Retrieve all recurring transactions for a specific budget."""
    result = await db.execute(
        text("""
            SELECT id, budget_id, category_id, title, amount, transaction_type,
                   frequency, day, active, created_at
            FROM recurring_transactions WHERE budget_id = :budget_id
        """),
        {"budget_id": budget_id}
    )
    rows = result.fetchall()
    return [RecurringTransaction(
        id=row.id,
        budget_id=row.budget_id,
        category_id=row.category_id,
        title=row.title,
        amount=row.amount,
        transaction_type=row.transaction_type,
        frequency=row.frequency,
        day=row.day,
        active=row.active,
        created_at=row.created_at,
    ) for row in rows]


@router.post("/budgets/{budget_id}/recurring", response_model=RecurringTransaction)
async def create_recurring_transaction(
    budget_id: str,
    payload: CreateRecurringTransaction,
    db: AsyncSession = Depends(get_db)
):
    """Create a new recurring transaction."""
    recurring_id = str(uuid4())
    now = datetime.now(timezone.utc).isoformat()

    await db.execute(
        text("""
            INSERT INTO recurring_transactions (id, budget_id, category_id, title, amount,
                                                 transaction_type, frequency, day, active, created_at)
            VALUES (:id, :budget_id, :category_id, :title, :amount,
                    :transaction_type, :frequency, :day, 1, :created_at)
        """),
        {
            "id": recurring_id,
            "budget_id": budget_id,
            "category_id": payload.category_id,
            "title": payload.title,
            "amount": payload.amount,
            "transaction_type": payload.transaction_type,
            "frequency": payload.frequency,
            "day": payload.day,
            "created_at": now,
        }
    )
    await db.commit()

    result = await db.execute(
        text("""
            SELECT id, budget_id, category_id, title, amount, transaction_type,
                   frequency, day, active, created_at
            FROM recurring_transactions WHERE id = :id
        """),
        {"id": recurring_id}
    )
    row = result.fetchone()
    return RecurringTransaction(
        id=row.id,
        budget_id=row.budget_id,
        category_id=row.category_id,
        title=row.title,
        amount=row.amount,
        transaction_type=row.transaction_type,
        frequency=row.frequency,
        day=row.day,
        active=row.active,
        created_at=row.created_at,
    )


@router.put("/recurring/{recurring_id}/toggle", response_model=RecurringTransaction)
async def toggle_recurring_transaction(recurring_id: str, db: AsyncSession = Depends(get_db)):
    """Toggle the active status of a recurring transaction."""
    result = await db.execute(
        text("SELECT active FROM recurring_transactions WHERE id = :id"),
        {"id": recurring_id}
    )
    row = result.fetchone()

    if not row:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Recurring transaction not found")

    new_active = 0 if row.active else 1

    await db.execute(
        text("UPDATE recurring_transactions SET active = :active WHERE id = :id"),
        {"active": new_active, "id": recurring_id}
    )
    await db.commit()

    result = await db.execute(
        text("""
            SELECT id, budget_id, category_id, title, amount, transaction_type,
                   frequency, day, active, created_at
            FROM recurring_transactions WHERE id = :id
        """),
        {"id": recurring_id}
    )
    row = result.fetchone()
    return RecurringTransaction(
        id=row.id,
        budget_id=row.budget_id,
        category_id=row.category_id,
        title=row.title,
        amount=row.amount,
        transaction_type=row.transaction_type,
        frequency=row.frequency,
        day=row.day,
        active=row.active,
        created_at=row.created_at,
    )


@router.delete("/recurring/{recurring_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_recurring_transaction(recurring_id: str, db: AsyncSession = Depends(get_db)):
    """Delete a recurring transaction."""
    await db.execute(text("DELETE FROM recurring_transactions WHERE id = :id"), {"id": recurring_id})
    await db.commit()


@router.post("/budgets/{budget_id}/recurring/process", response_model=list[Transaction])
async def process_recurring_transactions(budget_id: str, db: AsyncSession = Depends(get_db)):
    """
    Process recurring transactions and generate actual transactions.

    For each active recurring transaction, checks if transactions should have been
    created for the current month and creates them if missing.
    """
    from calendar import monthrange

    now = datetime.now(timezone.utc)
    current_year = now.year
    current_month = now.month
    current_day = now.day

    # Get first and last day of current month
    first_day_of_month = datetime(current_year, current_month, 1, tzinfo=timezone.utc)
    last_day_of_month = monthrange(current_year, current_month)[1]

    # Get all active recurring transactions for this budget
    result = await db.execute(
        text("""
            SELECT id, budget_id, category_id, title, amount, transaction_type,
                   frequency, day, active, created_at
            FROM recurring_transactions
            WHERE budget_id = :budget_id AND active = 1
        """),
        {"budget_id": budget_id}
    )
    recurring_rows = result.fetchall()

    created_transactions = []

    for rec in recurring_rows:
        dates_to_create = []

        if rec.frequency == 'monthly':
            # Monthly: create on the specified day (or last day if day > days in month)
            target_day = min(rec.day or 1, last_day_of_month)
            if target_day <= current_day:
                dates_to_create.append(f"{current_year}-{current_month:02d}-{target_day:02d}")

        elif rec.frequency == 'weekly':
            # Weekly: create for each week of the month that has passed
            # Find all occurrences of the recurring day this month
            rec_day_of_week = rec.day or 0  # 0 = Monday, 6 = Sunday

            for day in range(1, current_day + 1):
                date = datetime(current_year, current_month, day, tzinfo=timezone.utc)
                if date.weekday() == rec_day_of_week:
                    dates_to_create.append(f"{current_year}-{current_month:02d}-{day:02d}")

        elif rec.frequency == 'yearly':
            # Yearly: check if the day/month matches current month
            target_day = min(rec.day or 1, last_day_of_month)
            # For yearly, we only create if we're in the right month (assume created month)
            rec_created = datetime.fromisoformat(rec.created_at.replace('Z', '+00:00'))
            if rec_created.month == current_month and target_day <= current_day:
                dates_to_create.append(f"{current_year}-{current_month:02d}-{target_day:02d}")

        # Check which transactions already exist
        for date_str in dates_to_create:
            # Check if transaction already exists for this date and recurring template
            existing = await db.execute(
                text("""
                    SELECT id FROM transactions
                    WHERE budget_id = :budget_id
                    AND category_id = :category_id
                    AND title = :title
                    AND amount = :amount
                    AND date = :date
                    AND is_recurring = 1
                """),
                {
                    "budget_id": budget_id,
                    "category_id": rec.category_id,
                    "title": rec.title,
                    "amount": rec.amount,
                    "date": date_str,
                }
            )

            if existing.fetchone() is None:
                # Create the transaction
                transaction_id = str(uuid4())
                created_at = datetime.now(timezone.utc).isoformat()

                await db.execute(
                    text("""
                        INSERT INTO transactions (id, budget_id, category_id, title, amount,
                                                   transaction_type, date, comment, is_recurring, created_at)
                        VALUES (:id, :budget_id, :category_id, :title, :amount,
                                :transaction_type, :date, :comment, 1, :created_at)
                    """),
                    {
                        "id": transaction_id,
                        "budget_id": budget_id,
                        "category_id": rec.category_id,
                        "title": rec.title,
                        "amount": rec.amount,
                        "transaction_type": rec.transaction_type,
                        "date": date_str,
                        "comment": f"Auto-generated from recurring: {rec.title}",
                        "created_at": created_at,
                    }
                )

                created_transactions.append(Transaction(
                    id=transaction_id,
                    budget_id=budget_id,
                    category_id=rec.category_id,
                    title=rec.title,
                    amount=rec.amount,
                    transaction_type=rec.transaction_type,
                    date=date_str,
                    comment=f"Auto-generated from recurring: {rec.title}",
                    is_recurring=1,
                    created_at=created_at,
                ))

    await db.commit()
    return created_transactions
